db/test/versions/0056_test.js

```diff
@@ -0,0 +1,151 @@
+    onlineDowngrade: true,
+    createData: async client => {
+      // create the data including a task_queue_id that must be the
+      // combined identifier of provisioner_id/worker_type.
+      // this is a safe assumption as tested for the previous version.
+      await client.query(`
+        with gen as (
+          select generate_series(1, 99) as i
+        )
+        insert into tasks (task_id,
+                           provisioner_id,
+                           worker_type,
+                           scheduler_id,
+                           task_group_id,
+                           dependencies,
+                           requires,
+                           routes,
+                           priority,
+                           retries,
+                           created,
+                           deadline,
+                           expires,
+                           scopes,
+                           payload,
+                           metadata,
+                           tags,
+                           extra,
+                           retries_left,
+                           runs,
+                           taken_until,
+                           ever_resolved,
+                           task_queue_id)
+        select ${makeFieldsForCreation({ taskId: `'tid-' || gen.i`, workerType: `'wt-' || gen.i`, withDefaults: true })}
+        from gen`);
+    },
+    startCheck: async client => {
+      const res = await client.query('select task_id, task_queue_id from tasks');
+      const nextTaskId = res.rows.length + 1;
+      assert(nextTaskId >= 100, 'data was not created properly');
+      const tqids = res.rows.map(({ task_queue_id }) => task_queue_id).sort();
+      assert.deepEqual(tqids, _.range(1, nextTaskId).map(i => `pp/wt-${i}`).sort());
+
+      // check the schema
+      await helper.assertTableColumn('tasks', 'provisioner_id');
+      await helper.assertTableColumn('tasks', 'worker_type');
+      await helper.assertTableColumn('tasks', 'task_queue_id');
+    },
+    concurrentCheck: async client => {
+      // check that the inserted data looks as expected
+      let res = await client.query('select task_id, task_queue_id from tasks');
+      const nextTaskId = res.rows.length + 1;
+      const tids = res.rows.map(({ task_id }) => task_id).sort();
+      assert.deepEqual(tids, _.range(1, nextTaskId).map(i => `tid-${i}`).sort());
+      const tqids = res.rows.map(({ task_queue_id }) => task_queue_id).sort();
+      assert.deepEqual(tqids, _.range(1, nextTaskId).map(i => `pp/wt-${i}`).sort());
+
+      // check that get_task function works with items that may not yet have
+      // provisioner_id and worker_type during a downgrade
+      try {
+        res = await client.query(`select task_id from tasks 
+                                where worker_type is null or provisioner_id is null`);
+      } catch (err) {
+        if (err.code !== UNDEFINED_COLUMN) {
+          throw err;
+        }
```

Comment:
I'm not sure I understand what you are trying to say here. IIUC, `concurrentCheck` is indeed run before the downgradeScript, due to:
https://github.com/taskcluster/taskcluster/blob/06b8661b2741eee74d0c88912cb0920a35e7acf8/db/test/helper.js#L310-L316
and
https://github.com/taskcluster/taskcluster/blob/06b8661b2741eee74d0c88912cb0920a35e7acf8/db/test/helper.js#L354-L355

And that is why the try-catch is needed here, to only try to check `get_task` when those columns have already been created. The conditional here is only to ensure that we only omit the checks due to missing columns and not due to any other exception that it may occur (even if that's unlikely).

Reply:
You are absolutely right.  I tested this by deleting the conditional, when I should have replaced it with `throw err`.  Sorry!